Undefined names: 'vals', '_scatter', and 'rescatter'

[flake8](http://flake8.pycqa.org) testing of https://github.com/google/jax  on Python 3.7.1

$ __flake8 . --count --select=E901,E999,F821,F822,F823 --show-source --statistics__
```
./jax/interpreters/parallel.py:249:38: F821 undefined name 'vals'
      return call_primitive.bind(f, *vals, **params)
                                     ^
./jax/interpreters/parallel.py:372:10: F821 undefined name '_scatter'
  return _scatter(source, target, target_axis, name)
         ^
./jax/interpreters/parallel.py:422:9: F821 undefined name 'rescatter'
    x = rescatter(x, ydim, name)
        ^
3     F821 undefined name 'vals'
3
```
__E901,E999,F821,F822,F823__ are the "_showstopper_" [flake8](http://flake8.pycqa.org) issues that can halt the runtime with a SyntaxError, NameError, etc. These 5 are different from most other flake8 issues which are merely "style violations" -- useful for readability but they do not effect runtime safety.
* F821: undefined name `name`
* F822: undefined name `name` in `__all__`
* F823: local variable name referenced before assignment
* E901: SyntaxError or IndentationError
* E999: SyntaxError -- failed to compile a file into an Abstract Syntax Tree

--- a/jax/experimental/optimizers.py
+++ b/jax/experimental/optimizers.py
@@ -307,7 +307,7 @@ def rmsprop_momentum(step_size, gamma=0.9, eps=1e-8, momentum=0.9):
   Returns:
     An (init_fun, update_fun, get_params) triple.
   """
-  step_size = optimizers.make_schedule(step_size)
+  step_size = make_schedule(step_size)
   def init(x0):
     avg_sq_grad = np.zeros_like(x0)
     mom = np.zeros_like(x0)

--- a/jax/interpreters/pxla.py
+++ b/jax/interpreters/pxla.py
@@ -176,7 +176,7 @@ def remove_axis_from_aval(aval):
   elif isinstance(aval, ShapedArray):
     return ShapedArray(aval.shape[1:], aval.dtype)
   else:
-    raise TypeError(t)
+    raise TypeError(aval)
 
 
 def assign_shards_to_replicas(nrep, size):

--- a/jax/interpreters/xla.py
+++ b/jax/interpreters/xla.py
@@ -216,7 +216,7 @@ def _compile_jaxpr(jaxpr, device_assignment, axis_env, const_vals, *abstract_arg
   if axis_env.nreps > xb.device_count():
     msg = ("compiling computation that requires {} replicas, but only {} XLA "
            "devices are available")
-    raise ValueErrr(msg.format(axis_env.nreps, xb.device_count()))
+    raise ValueError(msg.format(axis_env.nreps, xb.device_count()))
   arg_shapes = list(map(xla_shape, abstract_args))
   built_c = _jaxpr_computation(jaxpr, axis_env, const_vals, (), *arg_shapes)
   result_shape = xla_shape_to_result_shape(built_c.GetReturnValueShape())

--- a/jax/lax/lax_parallel.py
+++ b/jax/lax/lax_parallel.py
@@ -635,7 +635,7 @@ def _broadcast_in_dim_papply_rule(name, size, vals, dims, shape,
             shape[dim], shape[out_dim]))
   sub_bdims = tuple(onp.delete(broadcast_dimensions, dim))
   sub_shape = tuple(onp.delete(shape, out_dim))
-  return broadcast_in_dim(operand, sub_shape, sub_bdims), out_dim
+  return lax.broadcast_in_dim(operand, sub_shape, sub_bdims), out_dim
 
 
 def _pad_papply_rule(name, size, vals, dims, padding_config):

--- a/jax/scipy/stats/bernoulli.py
+++ b/jax/scipy/stats/bernoulli.py
@@ -20,20 +20,20 @@ import numpy as onp
 import scipy.stats as osp_stats
 
 from ... import lax
-from ...numpy.lax_numpy import (_promote_args_like, _constant_like, _wraps,
-                                where, inf, logical_or)
+from ...numpy import lax_numpy as np
 from ..special import xlogy, xlog1py
 
 
-@_wraps(osp_stats.bernoulli.logpmf)
+@np._wraps(osp_stats.bernoulli.logpmf)
 def logpmf(k, p, loc=0):
-  k, p, loc = _promote_args_like(osp_stats.bernoulli.logpmf, k, p, loc)
-  zero = _constant_like(k, 0)
-  one = _constant_like(k, 1)
+  k, p, loc = np._promote_args_like(osp_stats.bernoulli.logpmf, k, p, loc)
+  zero = np._constant_like(k, 0)
+  one = np._constant_like(k, 1)
   x = lax.sub(k, loc)
   log_probs = xlogy(x, p) + xlog1py(lax.sub(one, x), -p)
-  return where(logical_or(lax.lt(x, zero), lax.gt(x, one)), -inf, log_probs)
+  return np.where(np.logical_or(lax.lt(x, zero), lax.gt(x, one)),
+                  -np.inf, log_probs)
 
-@_wraps(osp_stats.bernoulli.pmf)
+@np._wraps(osp_stats.bernoulli.pmf)
 def pmf(k, p, loc=0):
   return np.exp(pmf(k, p, loc))

--- a/jax/scipy/stats/dirichlet.py
+++ b/jax/scipy/stats/dirichlet.py
@@ -20,26 +20,26 @@ import numpy as onp
 import scipy.stats as osp_stats
 
 from ... import lax
-from ...numpy.lax_numpy import _promote_args_like, _constant_like, _wraps, all, sum
+from ...numpy import lax_numpy as np
 from ..special import gammaln, xlogy
 
 
 def _is_simplex(x):
-    x_sum = sum(x, axis=-1)
-    return all(x > 0, axis=-1) & (x_sum <= 1) & (x_sum > 1 - 1e-6)
+    x_sum = np.sum(x, axis=-1)
+    return np.all(x > 0, axis=-1) & (x_sum <= 1) & (x_sum > 1 - 1e-6)
 
 
-@_wraps(osp_stats.dirichlet.logpdf)
+@np._wraps(osp_stats.dirichlet.logpdf)
 def logpdf(x, alpha):
     args = (onp.ones((0,), lax.dtype(x)), onp.ones((1,), lax.dtype(alpha)))
     to_dtype = lax.dtype(osp_stats.dirichlet.logpdf(*args))
     x, alpha = [lax.convert_element_type(arg, to_dtype) for arg in (x, alpha)]
-    one = _constant_like(x, 1)
-    normalize_term = sum(gammaln(alpha), axis=-1) - gammaln(sum(alpha, axis=-1))
-    log_probs = lax.sub(sum(xlogy(lax.sub(alpha, one), x), axis=-1), normalize_term)
-    return where(_is_simplex(x), log_probs, -inf)
+    one = np._constant_like(x, 1)
+    normalize_term = np.sum(gammaln(alpha), axis=-1) - gammaln(np.sum(alpha, axis=-1))
+    log_probs = lax.sub(np.sum(xlogy(lax.sub(alpha, one), x), axis=-1), normalize_term)
+    return np.where(_is_simplex(x), log_probs, -np.inf)
 
 
-@_wraps(osp_stats.dirichlet.pdf)
+@np._wraps(osp_stats.dirichlet.pdf)
 def pdf(x, alpha):
   return lax.exp(logpdf(x, alpha))

--- a/tests/parallel_test.py
+++ b/tests/parallel_test.py
@@ -137,7 +137,7 @@ class ParallelizeTest(jtu.JaxTestCase):
 
   def dedup(self, arr, expected_rank):
     if arr.ndim == expected_rank + 1:
-      for i in xrange(arr.shape[0] - 1):
+      for i in range(arr.shape[0] - 1):
         self.assertAllClose(arr[i], arr[i + 1], check_dtypes=True)
       return arr[0]
     else:

--- a/tests/random_test.py
+++ b/tests/random_test.py
@@ -28,6 +28,7 @@ import scipy.stats
 
 from jax import api
 from jax import lax
+from jax import numpy as np
 from jax import random
 from jax import test_util as jtu
 from jax.interpreters import xla

